src/object/srv_obj.c

```diff
@@ -720,85 +779,229 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
+	if (update && !(cont_hdl->sch_capas & DAOS_COO_RW)) {
 		D_ERROR("cont "DF_UUID" sch_capas "DF_U64", "
 			"NO_PERM to update.\n",
 			DP_UUID(orw->orw_co_uuid), cont_hdl->sch_capas);
-		D_GOTO(out, rc = -DER_NO_PERM);
+		D_GOTO(reply, rc = -DER_NO_PERM);
 	}
 
 	D_ASSERT(cont_hdl->sch_pool != NULL);
 	map_ver = cont_hdl->sch_pool->spc_map_version;
+
+	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
+		DF_U64", pool ver %u/%u with "DF_DTI".\n",
+		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
+		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
+		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
+
 	if (orw->orw_map_ver < map_ver) {
 		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
 			orw->orw_map_ver, map_ver);
-		if (update && dispatch)
-			D_GOTO(out, rc = -DER_STALE);
+		if (update)
+			D_GOTO(reply, rc = -DER_STALE);
+
+		/* It is harmless if fetch with old pool map version. */
+	} else if (orw->orw_map_ver > map_ver) {
+		/* XXX: Client (or leader replica) has newer pool map than
+		 *	current replica. Two possibile cases:
+		 *
+		 *	1. The current replica was the old leader if with
+		 *	   the old pool map version. According to current
+		 *	   leader election algorithm, it is still the new
+		 *	   leader with the new pool map version. Since no
+		 *	   leader switch, the unmatched pool version will
+		 *	   not affect DTX related availability check.
+		 *
+		 *	2. The current replica was NOT the old leader if
+		 *	   with the old pool map version. But it becomes
+		 *	   the new leader with the new pool map version.
+		 *	   In the subsequent modificaiton, it may hit
+		 *	   some 'prepared' DTX when make availability
+		 *	   check, it will return -DER_INPROGRESS that
+		 *	   will cause client to retry. It is possible
+		 *	   that the pool map version event arrives at
+		 *	   this server during the client retry. It is
+		 *	   inefficient, but harmless.
+		 */
+	}
+
+	if (!update) {
+		rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
+					  &ioh, false, map_ver, dth);
+		if (rc != 0)
+			D_ERROR(DF_UOID": rw_local_hdlr (fetch) failed %d.\n",
+				DP_UOID(orw->orw_oid), rc);
+
+		D_GOTO(reply, rc);
 	}
 
-	/* dispatch to other tgts when needed */
+	if (!daos_oc_echo_type(daos_obj_id2class(orw->orw_oid.id_pub))) {
+		/* Check DTX configuration. */
+		rc = ds_obj_check_dtx_config(orw->orw_flags, &orw->orw_dti);
+		if (rc != 0)
+			D_GOTO(reply, rc);
+
+		/* Handle resend. */
+		if (srv_enable_dtx && orw->orw_flags & ORF_RESEND) {
+			rc = dtx_handle_resend(cont->sc_hdl, &orw->orw_oid,
+					       &orw->orw_dti,
+					       orw->orw_dkey_hash,
+					       srv_start_time, false);
+			if (rc == -DER_ALREADY)
+				D_GOTO(reply, rc = 0);
+			if (rc == 0)
+				resend = true;
+			else if (rc == -DER_NONEXIST)
+				rc = 0;
+			else
+				D_GOTO(reply, rc);
+		}
+	}
+
+	if (srv_enable_dtx && (!resend || dispatch)) {
```

Comment:
in the case of prepared & resend and dispatch, still need to do the dtx_begin()?
or should it be "if (srv_enable_dtx && !resend)"?

Reply:
The leader may have done the modification, but it does not know whether non-leader replicas have done or not, so even if the leader know it is resent case, it still needs to dispatch the RPC.